### PR TITLE
[Testcase Refactoring] Decouple test_fsdp_core from hard-coded devices and label ACCELERATOR classes

### PR DESCRIPTION
**Pull request by twthighd**
## Summary

Label the five `FSDPTest` subclasses in `test_fsdp_core.py` as `ACCELERATOR`, replace the hard-coded `only_for=("cuda","hpu","xpu")` whitelist with `except_for=("cpu",)`, migrate all test methods and helpers from the module-level `device_type` constant to the injected `device` parameter (type-only), and add a `_device_sleep` helper in `common_fsdp.py` to replace hard-coded CUDA/HPU/XPU delay branches. Review-feedback updates: the `@unittest.skipIf(TEST_HPU, ...)` guard is replaced with the variant-level `@skipHPUIf(True, ...)`, and the `device` parameter is moved to the second position in all test method signatures.

## Changes

- **`test/distributed/fsdp/test_fsdp_core.py`**:
  - Add `hw_classification = HardwareClassification.ACCELERATOR` to `TestParityWithDDP`, `TestParamInit`, `TestHooks`, `TestNoGrad`, `TestAutograd`. Place the label after the class docstring in `TestParityWithDDP` so the docstring remains the first statement and `__doc__` is preserved.
  - Replace `devices = ("cuda", "hpu", "xpu")` + `only_for=devices` with `except_for=("cpu",)` in all five `instantiate_device_type_tests` calls, so any accelerator registered via `torch.accelerator` (including PrivateUse1 out-of-tree backends) is admitted.
  - Migrate all 14 `test_*` methods and the `_test_pre_backward_hook_registration` / `_test_unshard_params_as_tensors` helpers from the module-level `device_type = torch.device(get_devtype())` constant to the injected `device` parameter. For FSDP `device_id`, `nn.Linear(device=...)`, `torch.randn(device=...)`, `.to(...)`, and `get_input(...)`, use `torch.device(device).type` (type-only, index stripped) to preserve FSDP's per-rank current-device resolution semantics. Delete the module-level `device_type` and the `get_devtype` import.
  - Replace the environment-level `@unittest.skipIf(TEST_HPU, ...)` on `test_mixture_of_experts_with_delay_before_free` with the variant-level `@skipHPUIf(True, ...)` from `common_device_type`, so only the HPU device variant is skipped instead of all variants on an HPU host; remove the now-unused `TEST_HPU` import (review feedback).
  - Move the `device` parameter to the second position (right after `self`) in all test method signatures, matching the convention of the recent decoupling PRs (review feedback). This is signature-only — parameter injection is keyword-based, so runtime behavior is unchanged.
- **`torch/testing/_internal/common_fsdp.py`**:
  - Add a `_device_sleep(ms)` helper that resolves the current accelerator via `torch.get_device_module` and uses its device-side `_sleep` when available, falling back to host-side `time.sleep` otherwise. Replace the three hard-coded CUDA/HPU/XPU delay branches in `ModuleWithDelay` and `MixtureOfExperts` with calls to it.

## Motivation

The five classes run FSDP device-tensor paths guarded by `@skip_if_lt_x_gpu(2)` and are not CPU-runnable, so `ACCELERATOR` is the correct classification. The original `only_for=("cuda","hpu","xpu")` whitelist excluded PrivateUse1-based out-of-tree backends; `except_for=("cpu",)` lets the framework auto-generate all available device variants. The `device` parameter migration is required because `instantiate_device_type_tests` generates per-device variants whose `cls.device_type` differs, and a module-level constant cannot follow. The type-only form (`torch.device(device).type`) is required for FSDP `device_id` and tensor creation because the injected `device` is an indexed string (e.g. `"cuda:0"`) resolved via `CUDATestBase.get_primary_device`; passing it bare forces all ranks onto device 0, degrading the multi-device test and producing numerical mismatches from the FSDP auto-move path.

## Test Plan

- Verified end-to-end on an 8-device accelerator backend (with the `common_fsdp.py` accelerator branch from #187650 applied):

```bash
python test/distributed/fsdp/test_fsdp_core.py
# Ran 60 tests in 1010.744s — OK (skipped=6)
```

- `lintrunner` reports no lint issues.

## Notes

- `skip_if_lt_x_gpu` device decoupling (letting that helper recognize out-of-tree accelerators) is tracked by pytorch/pytorch#187650 — an independent framework-level change to `common_distributed.py`, orthogonal to this PR. This PR keeps `@skip_if_lt_x_gpu(2)` unchanged.
- For FSDP tests using the `FSDPTest` / `FSDPTestContinuous` base classes to truly run on an out-of-tree accelerator (rather than falling back to a CPU/gloo path), the `common_fsdp.py` device/backend selection also needs the accelerator branch that #187650 adds. That is a framework-level change tracked by #187650, separate from this PR.
- This is part of the test case refactoring initiative tracked in #185590.

### CODE REVIEW

**Comment by twthighd** [comment]
@wjlFlyer @pjyFight @FuDdd @fffrog @lyriexs666 @jishuangfeng @JamesD18 @JiasenTian

File: test/distributed/fsdp/test_fsdp_core.py
Comment (wjlFlyer) [inline test/distributed/fsdp/test_fsdp_core.py]: `TEST_HPU` should be removed.

File: test/distributed/fsdp/test_fsdp_core.py [hunk 1]
```diff
@@ -264,20 +277,22 @@ def test_mixture_of_experts_with_delay_before_free(
...
         )
 
 
 class TestParamInit(FSDPTest):
+    hw_classification = HardwareClassification.ACCELERATOR
+
     @skip_if_lt_x_gpu(2)
     @parametrize("mixed_precision", [True, False])
-    def test_param_change_after_init(self, mixed_precision):
+    def test_param_change_after_init(self, mixed_precision, device):
```
Comment (wjlFlyer) [inline test/distributed/fsdp/test_fsdp_core.py]: Should `device` in the second position of all parameters? Please confirm that.

File: test/distributed/fsdp/test_fsdp_core.py
Comment (twthighd) [reply test/distributed/fsdp/test_fsdp_core.py]: Agreed — replaced with the variant-level @skipHPUIf(True, ...) from common_device_type so only the HPU variant is skipped; TEST_HPU import removed. Thanks for the review!

File: test/distributed/fsdp/test_fsdp_core.py (hunk 1 above)
Comment (twthighd) [reply test/distributed/fsdp/test_fsdp_core.py]: Confirmed — moving device to the second position (right after self) in all 14 signatures, matching the convention in the recent decoupling PRs.

File: test/distributed/fsdp/test_fsdp_core.py [hunk 2]
```diff
@@ -120,12 +124,14 @@ def test_nested_wrapped_model(
             FSDPInitMode.RECURSIVE,
             cpu_offload=cpu_offload,
             sharding_strategy=sharding_strategy,
+            device_id=torch.device(device).type,
```
Comment (wjlFlyer) [inline test/distributed/fsdp/test_fsdp_core.py:127]: `device` parameter is str type, so could we use `device_id=device` rather than `device_id=torch.device(device).type`?

Comment (twthighd) [reply test/distributed/fsdp/test_fsdp_core.py:127]: Thanks for the review! The injected device is an indexed string ("cuda:0"-style — it comes from get_primary_device(), which returns f"{device_type}:{current_device()}"), not a type-only one. Passing it directly as device_id pins every rank's FSDP shards to device 0, since the index is honored as-is — that silently degenerates the multi-device verification into single-device. torch.device(device).type strips the index, so FSDP falls back to current_device() per rank (which set_device_index sets up per rank on CUDA/XPU) and each rank stays on its own device.

**Review by wjlFlyer — APPROVED** [review]
